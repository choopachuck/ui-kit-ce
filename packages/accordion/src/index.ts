export * from './Accordion'
export * from './AccordionItem'
export type {
  AccordionIconProps,
  AccordionComponentsConfig,
} from './components'
