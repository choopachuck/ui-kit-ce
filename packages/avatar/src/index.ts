export * from './Avatar'
export * from './AvatarGroup'
export * from './types'
export type {
  AvatarGroupExtraProps,
  AvatarGroupComponentsConfig,
} from './components'
export * from './classes'
export * from './constants'
