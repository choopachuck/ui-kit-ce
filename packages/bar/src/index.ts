export * from './Bar'
export * from './BarButton'
export * from './BarDate'
export * from './BarDivider'
export * from './BarDropdown'
export * from './BarDropdownItem'
export * from './BarItem'
export * from './BarMenuItem'
export * from './BarSearch'
export * from './BarSelect'
export * from './constants'
export * from './interfaces'
export * from './SubBar'
