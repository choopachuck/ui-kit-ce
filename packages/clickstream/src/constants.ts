export const CLICK_STREAM_EVENTS = ['click', 'change'] as const

export const CLICK_STREAM_INACTIVITY_TIME = 30000

export const CLICK_STREAM_BATCH_SIZE = 10

// Нотация нарушена для обратной совместимости с @sbol/clickstream-agent
export const CONNECTION_TYPES = {
  bluetooth: 'BLUETOOTH',
  cellular: 'CELLULAR',
  ethernet: 'ETHERNET',
  wifi: 'WIFI',
  mixed: 'UNKNOWN',
  none: 'UNKNOWN',
  wimax: 'UNKNOWN',
  other: 'UNKNOWN',
  unknown: 'UNKNOWN',
} as const

export const CLICK_STREAM_GENERIC_ATTRIBUTES = {
  OFF: 'data-click-stream-off',
} as const

export const ALIGNMENT = 10

export const MINUTES_IN_HOUR = 60

export const MS_IN_SECOND = 1000

export const SECONDS_IN_MINUTE = 60

export const MS_IN_MINUTE = SECONDS_IN_MINUTE * MS_IN_SECOND
