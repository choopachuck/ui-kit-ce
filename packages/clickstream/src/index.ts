export * from './ClickStreamProvider'
export type {
  ClickStreamBaseData,
  ClickStreamContextProps,
  ClickStreamElementProperty,
  ClickStreamElementType,
  ClickStreamEventBatch,
  ClickStreamEventBatchItem,
  ClickStreamEventData,
  ClickStreamEventMeta,
  ClickStreamProviderProps,
} from './types'
