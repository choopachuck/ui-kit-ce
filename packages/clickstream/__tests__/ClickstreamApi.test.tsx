/*--------------------------------------------------------------------------------------------------
 * ----------------------- Тестирование API компонента ClickStreamProvider -------------------------
 * -------------------------------------------------------------------------------------------------
 */

import * as React from 'react'
import { render, waitFor, fireEvent } from '@testing-library/react'
import {
  ClickStreamProvider,
  ClickStreamRef,
  ClickStreamProviderProps,
} from '../src'
import * as ClickStreamUtils from '../src/utils'
import { mockGlobals } from './helpers'

jest.useFakeTimers('modern').setSystemTime(new Date('2020-01-01'))
jest
  .spyOn(ClickStreamUtils, 'getISOTime')
  .mockImplementation(() => '2020-01-01T00:00:00.078-00:00')

const ClickStreamProviderRefTest: React.FC<ClickStreamProviderProps> = (
  props
) => {
  const ref = React.useRef<ClickStreamRef>(null)

  const handleButtonClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    ref?.current?.dispatchEvent(event.nativeEvent, 'customEvent')
  }

  return (
    <ClickStreamProvider {...props} ref={ref}>
      <button onClick={handleButtonClick}>button</button>
    </ClickStreamProvider>
  )
}

beforeAll(() => {
  mockGlobals()
})

it('onInit callback works correctly', async () => {
  const handleInit = jest.fn()
  render(<ClickStreamProvider onInit={handleInit} />)

  await waitFor(() => expect(handleInit).toHaveBeenCalledTimes(1))
  expect(handleInit).toMatchSnapshot()
})

it('onSendEvent callback works correctly with generic batching', async () => {
  const actionCount = 10
  const handleSendEvent = jest.fn()
  const { getByRole } = render(
    <ClickStreamProvider onSendEvent={handleSendEvent}>
      <button id="custom-second-button" aria-label="custom button">
        0
      </button>
    </ClickStreamProvider>
  )

  const button = getByRole('button')

  for (let i = 0; i < actionCount; i++) {
    fireEvent.click(button)
  }

  await waitFor(() => expect(handleSendEvent).toHaveBeenCalledTimes(1))
  expect(handleSendEvent).toMatchSnapshot()
})

it('onSendEvent callback works correctly with custom batching', async () => {
  const batchSize = 5
  const actionCount = 10
  const handleSendEvent = jest.fn()
  const { getByRole } = render(
    <ClickStreamProvider batchSize={batchSize} onSendEvent={handleSendEvent}>
      <button id="third-button" aria-label="custom button">
        0
      </button>
    </ClickStreamProvider>
  )

  const button = getByRole('button')

  for (let i = 0; i < actionCount; i++) {
    fireEvent.click(button)
  }

  await waitFor(() => expect(handleSendEvent).toHaveBeenCalledTimes(2))
  expect(handleSendEvent).toMatchSnapshot()
})

it('onSendEvent callback works correctly at the end of the timer', async () => {
  const actionCount = 9
  const inactivityTime = 2000
  const handleSendEvent = jest.fn()
  const { getByRole } = render(
    <ClickStreamProvider
      inactivityTime={inactivityTime}
      onSendEvent={handleSendEvent}
    >
      <button id="personal-button" aria-label="custom button">
        0
      </button>
    </ClickStreamProvider>
  )

  const button = getByRole('button')

  for (let i = 0; i < actionCount; i++) {
    fireEvent.click(button)
  }

  // Добавляем небольшой разброс в 0.5 секунд, чтобы таймаут теста смог дождаться таймаута кликстрима
  await waitFor(() => expect(handleSendEvent).toHaveBeenCalledTimes(1), {
    timeout: inactivityTime + 500,
  })
  expect(handleSendEvent).toMatchSnapshot()
})

it('ref has all nessesary imperative handles', () => {
  const ref = jest.fn()
  render(<ClickStreamProvider ref={ref} />)

  expect(ref).toMatchSnapshot()
})

it('dispatch event from ref works correctly', async () => {
  const handleBatch = jest.fn()
  const { getByRole } = render(
    <ClickStreamProviderRefTest onBatch={handleBatch} />
  )

  fireEvent.click(getByRole('button'))

  await waitFor(() => expect(handleBatch).toHaveBeenCalledTimes(2))
  expect(handleBatch).toMatchSnapshot()
})

it('disableAutoSendEvent property prevents all automatic events', async () => {
  const handleSendEvent = jest.fn()
  const handleBatch = jest.fn()
  const actionCount = 10
  const { getByRole } = render(
    <ClickStreamProvider
      disableAutoSendEvent
      onSendEvent={handleSendEvent}
      onBatch={handleBatch}
    >
      <button id="secondary-button" aria-label="custom button">
        0
      </button>
    </ClickStreamProvider>
  )

  const button = getByRole('button')

  for (let i = 0; i < actionCount; i++) {
    fireEvent.click(button)
  }

  await waitFor(() => expect(handleSendEvent).toHaveBeenCalledTimes(0))
  await waitFor(() => expect(handleBatch).toHaveBeenCalledTimes(0))
})

it('data-click-stream-off property in child element prevents automatic event', async () => {
  const handleBatch = jest.fn()

  const { getByText } = render(
    <ClickStreamProvider onBatch={handleBatch}>
      <button data-click-stream-off id="local-button">
        button01
      </button>
      <button id="button02">button02</button>
    </ClickStreamProvider>
  )

  getByText('button01').click()
  getByText('button02').click()

  await waitFor(() => expect(handleBatch).toHaveBeenCalledTimes(1))
})

it('custom formatEventData works correctly', async () => {
  const formatEventData = jest.fn().mockImplementation(() => ({
    foo: '1',
    baar: '2',
  }))
  const { getByRole } = render(
    <ClickStreamProvider formatEventData={formatEventData}>
      <button id="button" aria-label="custom button">
        0
      </button>
    </ClickStreamProvider>
  )

  fireEvent.click(getByRole('button'))

  await waitFor(() => expect(formatEventData).toHaveBeenCalledTimes(1))
  expect(formatEventData).toMatchSnapshot()
})
