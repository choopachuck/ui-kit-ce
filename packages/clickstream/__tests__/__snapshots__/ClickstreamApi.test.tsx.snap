// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom formatEventData works correctly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      MouseEvent {
        "isTrusted": false,
      },
      "click",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "baar": "2",
        "foo": "1",
      },
    },
  ],
}
`;

exports[`dispatch event from ref works correctly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      MouseEvent {
        "isTrusted": false,
      },
      Object {
        "batteryLevel": "0",
        "connectionType": "UNKNOWN",
        "elementAttributes": Array [],
        "elementEvent": "customEvent",
        "elementTag": "button",
        "elementType": "submit",
        "pageName": "",
        "timeStamp": "2020-01-01T00:00:00.078-00:00",
        "url": "http://localhost/",
        "value": "button",
      },
      Array [
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [],
            "elementEvent": "customEvent",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
      ],
    ],
    Array [
      MouseEvent {
        "isTrusted": false,
      },
      Object {
        "batteryLevel": "0",
        "connectionType": "UNKNOWN",
        "elementAttributes": Array [],
        "elementEvent": "click",
        "elementTag": "button",
        "elementType": "submit",
        "pageName": "",
        "timeStamp": "2020-01-01T00:00:00.078-00:00",
        "url": "http://localhost/",
        "value": "button",
      },
      Array [
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [],
            "elementEvent": "customEvent",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
      ],
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`onInit callback works correctly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "batteryLevel": "0",
        "connectionType": "UNKNOWN",
        "pageName": "",
        "timeStamp": "2020-01-01T00:00:00.078-00:00",
        "url": "http://localhost/",
      },
      Object {
        "applicationLanguage": "ru",
        "browser": "Mozilla/5.0 (linux) AppleWebKit/537.36 (KHTML, like Gecko) jsdom/16.7.0",
        "platform": "WEB",
        "screenSize": "0x0",
        "systemLanguage": "ru",
        "timeStamp": "2020-01-01T00:00:00.078-00:00",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`onSendEvent callback works correctly at the end of the timer 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "personal-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "personal-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "personal-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "personal-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "personal-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "personal-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "personal-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "personal-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "personal-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
      ],
      Object {
        "applicationLanguage": "ru",
        "browser": "Mozilla/5.0 (linux) AppleWebKit/537.36 (KHTML, like Gecko) jsdom/16.7.0",
        "platform": "WEB",
        "screenSize": "0x0",
        "systemLanguage": "ru",
        "timeStamp": "2020-01-01T00:00:00.078-00:00",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`onSendEvent callback works correctly with custom batching 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "third-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "third-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "third-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "third-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "third-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
      ],
      Object {
        "applicationLanguage": "ru",
        "browser": "Mozilla/5.0 (linux) AppleWebKit/537.36 (KHTML, like Gecko) jsdom/16.7.0",
        "platform": "WEB",
        "screenSize": "0x0",
        "systemLanguage": "ru",
        "timeStamp": "2020-01-01T00:00:00.078-00:00",
      },
    ],
    Array [
      Array [
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "third-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "third-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "third-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "third-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "third-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
      ],
      Object {
        "applicationLanguage": "ru",
        "browser": "Mozilla/5.0 (linux) AppleWebKit/537.36 (KHTML, like Gecko) jsdom/16.7.0",
        "platform": "WEB",
        "screenSize": "0x0",
        "systemLanguage": "ru",
        "timeStamp": "2020-01-01T00:00:00.078-00:00",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`onSendEvent callback works correctly with generic batching 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "custom-second-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "custom-second-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "custom-second-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "custom-second-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "custom-second-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "custom-second-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "custom-second-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "custom-second-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "custom-second-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
        Object {
          "data": Object {
            "batteryLevel": "0",
            "connectionType": "UNKNOWN",
            "elementAttributes": Array [
              Object {
                "key": "id",
                "value": "custom-second-button",
              },
              Object {
                "key": "aria-label",
                "value": "custom button",
              },
            ],
            "elementEvent": "click",
            "elementTag": "button",
            "elementType": "submit",
            "pageName": "",
            "timeStamp": "2020-01-01T00:00:00.078-00:00",
            "url": "http://localhost/",
            "value": "custom button",
          },
          "event": MouseEvent {
            "isTrusted": false,
          },
        },
      ],
      Object {
        "applicationLanguage": "ru",
        "browser": "Mozilla/5.0 (linux) AppleWebKit/537.36 (KHTML, like Gecko) jsdom/16.7.0",
        "platform": "WEB",
        "screenSize": "0x0",
        "systemLanguage": "ru",
        "timeStamp": "2020-01-01T00:00:00.078-00:00",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`ref has all nessesary imperative handles 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "dispatchEvent": [Function],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
